File: bonus-typescript/src/research-assistant.ts
/**
 * Research Assistant Agent - TypeScript Implementation
 * Microsoft Agent Framework Series - Part 3 Equivalent
 */

import OpenAI from 'openai';
import { DefaultAzureCredential, getBearerTokenProvider } from '@azure/identity';
import * as dotenv from 'dotenv';

dotenv.config();

// Tool definitions
const tools: OpenAI.ChatCompletionTool[] = [
    {
        type: 'function',
        function: {
            name: 'search_web',
            description: 'Search the web for information on a topic',
            parameters: {
                type: 'object',
                properties: {
                    query: { type: 'string', description: 'The search query' },
                    num_results: { type: 'number', description: 'Number of results', default: 5 }
                },
                required: ['query']
            }
        }
    },
    {
        type: 'function',
        function: {
            name: 'summarize_text',
            description: 'Create a concise summary of the provided content',
            parameters: {
                type: 'object',
                properties: {
                    content: { type: 'string', description: 'Content to summarize' },
                    max_words: { type: 'number', description: 'Max words', default: 100 }
                },
                required: ['content']
            }
        }
    },
    {
        type: 'function',
        function: {
            name: 'get_date',
            description: 'Get the current date',
            parameters: { type: 'object', properties: {} }
        }
    }
];

// Tool implementations
function searchWeb(query: string, numResults: number = 5): string {
    console.log(`Searching web for: ${query}`);
    return `Top ${numResults} results for '${query}':
    1. [Enterprise AI Guide] - Microsoft Research
    2. [Agentic AI Patterns] - Harvard Business Review
    3. [Building AI Agents] - Azure AI Blog`;
}

function summarizeText(content: string, maxWords: number = 100): string {
    const words = content.split(' ');
    if (words.length <= maxWords) return content;
    return words.slice(0, maxWords).join(' ') + '...';
}

function getDate(): string {
    return new Date().toLocaleDateString('en-US', {
        year: 'numeric', month: 'long', day: 'numeric'
    });
}

// Execute tool calls
function executeTool(name: string, args: Record<string, unknown>): string {
    switch (name) {
        case 'search_web':
            return searchWeb(args.query as string, args.num_results as number);
        case 'summarize_text':
            return summarizeText(args.content as string, args.max_words as number);
        case 'get_date':
            return getDate();
        default:
            return `Unknown tool: ${name}`;
    }
}

// Agent class
class ResearchAssistant {
    private client: OpenAI;
    private messages: OpenAI.ChatCompletionMessageParam[] = [];
    private systemPrompt: string;

    constructor() {
        const credential = new DefaultAzureCredential();
        const scope = 'https://cognitiveservices.azure.com/.default';
        const azureADTokenProvider = getBearerTokenProvider(credential, scope);

        this.client = new OpenAI({
            apiKey: '', // Not needed with Azure AD
            baseURL: `${process.env.AZURE_OPENAI_ENDPOINT}/openai/deployments/${process.env.AZURE_OPENAI_DEPLOYMENT_NAME}`,
            defaultQuery: { 'api-version': '2024-08-01-preview' },
            defaultHeaders: { 'api-key': '' },
        });

        this.systemPrompt = `You are an expert research assistant. Your capabilities:
      1. Search the web for current information
      2. Summarize complex content concisely
      3. Give accurate, well-sourced answers
      Always cite your sources and use tools to find information.`;

        this.messages.push({ role: 'system', content: this.systemPrompt });
    }

    async run(userMessage: string): Promise<string> {
        this.messages.push({ role: 'user', content: userMessage });

        let response = await this.client.chat.completions.create({
            model: process.env.AZURE_OPENAI_DEPLOYMENT_NAME!,
            messages: this.messages,
            tools: tools,
            tool_choice: 'auto'
        });

        let assistantMessage = response.choices[0].message;

        // Handle tool calls
        while (assistantMessage.tool_calls && assistantMessage.tool_calls.length > 0) {
            this.messages.push(assistantMessage);

            for (const toolCall of assistantMessage.tool_calls) {
                const args = JSON.parse(toolCall.function.arguments);
                const result = executeTool(toolCall.function.name, args);

                this.messages.push({
                    role: 'tool',
                    tool_call_id: toolCall.id,
                    content: result
                });
            }

            response = await this.client.chat.completions.create({
                model: process.env.AZURE_OPENAI_DEPLOYMENT_NAME!,
                messages: this.messages,
                tools: tools,
                tool_choice: 'auto'
            });

            assistantMessage = response.choices[0].message;
        }

        const finalContent = assistantMessage.content || '';
        this.messages.push({ role: 'assistant', content: finalContent });

        return finalContent;
    }

    clearHistory(): void {
        this.messages = [{ role: 'system', content: this.systemPrompt }];
    }
}

// Main
async function main() {
    console.log('='.repeat(60));
    console.log('  🔬 Research Assistant v1.0 (TypeScript)');
    console.log('='.repeat(60));

    const assistant = new ResearchAssistant();

    // Example conversation
    const questions = [
        "What are the key trends in AI agents for 2025?",
        "Summarize the main benefits of agentic AI"
    ];

    for (const question of questions) {
        console.log(`\n📝 You: ${question}`);
        const response = await assistant.run(question);
        console.log(`\n🤖 Assistant: ${response}`);
    }
}

main().catch(console.error);

File: bonus-go/research_assistant.go
// Research Assistant Agent - Go Implementation
// Microsoft Agent Framework Series - Part 3 Equivalent

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"github.com/joho/godotenv"
	openai "github.com/sashabaranov/go-openai"
)

// Tool definitions
type Tool struct {
	Name        string
	Description string
	Execute     func(args map[string]interface{}) string
}

var tools = map[string]Tool{
	"search_web": {
		Name:        "search_web",
		Description: "Search the web for information",
		Execute: func(args map[string]interface{}) string {
			query := args["query"].(string)
			numResults := 5
			if n, ok := args["num_results"].(float64); ok {
				numResults = int(n)
			}
			log.Printf("Searching web for: %s", query)
			return fmt.Sprintf(`Top %d results for '%s':
1. [Enterprise AI Guide] - Microsoft Research
2. [Agentic AI Patterns] - Harvard Business Review
3. [Building AI Agents] - Azure AI Blog`, numResults, query)
		},
	},
	"summarize_text": {
		Name:        "summarize_text",
		Description: "Create a concise summary",
		Execute: func(args map[string]interface{}) string {
			content := args["content"].(string)
			maxWords := 100
			if n, ok := args["max_words"].(float64); ok {
				maxWords = int(n)
			}
			words := strings.Fields(content)
			if len(words) <= maxWords {
				return content
			}
			return strings.Join(words[:maxWords], " ") + "..."
		},
	},
	"get_date": {
		Name:        "get_date",
		Description: "Get the current date",
		Execute: func(args map[string]interface{}) string {
			return time.Now().Format("January 2, 2006")
		},
	},
}

// OpenAI tool definitions
func getOpenAITools() []openai.Tool {
	return []openai.Tool{
		{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name:        "search_web",
				Description: "Search the web for information on a topic",
				Parameters: map[string]interface{}{
					"type": "object",
					"properties": map[string]interface{}{
						"query":       map[string]string{"type": "string", "description": "Search query"},
						"num_results": map[string]interface{}{"type": "number", "description": "Number of results"},
					},
					"required": []string{"query"},
				},
			},
		},
		{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name:        "summarize_text",
				Description: "Create a concise summary of content",
				Parameters: map[string]interface{}{
					"type": "object",
					"properties": map[string]interface{}{
						"content":   map[string]string{"type": "string", "description": "Content to summarize"},
						"max_words": map[string]interface{}{"type": "number", "description": "Max words"},
					},
					"required": []string{"content"},
				},
			},
		},
		{
			Type: openai.ToolTypeFunction,
			Function: &openai.FunctionDefinition{
				Name:        "get_date",
				Description: "Get the current date",
				Parameters: map[string]interface{}{
					"type":       "object",
					"properties": map[string]interface{}{},
				},
			},
		},
	}
}

// ResearchAssistant agent
type ResearchAssistant struct {
	client   *openai.Client
	messages []openai.ChatCompletionMessage
}

func NewResearchAssistant() *ResearchAssistant {
	config := openai.DefaultAzureConfig(
		os.Getenv("AZURE_OPENAI_API_KEY"),
		os.Getenv("AZURE_OPENAI_ENDPOINT"),
	)
	config.APIVersion = "2024-08-01-preview"

	return &ResearchAssistant{
		client: openai.NewClientWithConfig(config),
		messages: []openai.ChatCompletionMessage{
			{
				Role: openai.ChatMessageRoleSystem,
				Content: `You are an expert research assistant. Your capabilities:
1. Search the web for current information
2. Summarize complex content concisely
3. Give accurate, well-sourced answers
Always cite your sources and use tools to find information.`,
			},
		},
	}
}

func (ra *ResearchAssistant) Run(ctx context.Context, userMessage string) (string, error) {
	ra.messages = append(ra.messages, openai.ChatCompletionMessage{
		Role:    openai.ChatMessageRoleUser,
		Content: userMessage,
	})

	for {
		resp, err := ra.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
			Model:    os.Getenv("AZURE_OPENAI_DEPLOYMENT_NAME"),
			Messages: ra.messages,
			Tools:    getOpenAITools(),
		})
		if err != nil {
			return "", err
		}

		assistantMsg := resp.Choices[0].Message

		// Check for tool calls
		if len(assistantMsg.ToolCalls) == 0 {
			ra.messages = append(ra.messages, assistantMsg)
			return assistantMsg.Content, nil
		}

		// Execute tool calls
		ra.messages = append(ra.messages, assistantMsg)

		for _, toolCall := range assistantMsg.ToolCalls {
			var args map[string]interface{}
			json.Unmarshal([]byte(toolCall.Function.Arguments), &args)

			tool := tools[toolCall.Function.Name]
			result := tool.Execute(args)

			ra.messages = append(ra.messages, openai.ChatCompletionMessage{
				Role:       openai.ChatMessageRoleTool,
				Content:    result,
				ToolCallID: toolCall.ID,
			})
		}
	}
}

func main() {
	godotenv.Load()

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("  🔬 Research Assistant v1.0 (Go)")
	fmt.Println(strings.Repeat("=", 60))

	assistant := NewResearchAssistant()
	ctx := context.Background()

	questions := []string{
		"What are the key trends in AI agents for 2025?",
		"Summarize the main benefits of agentic AI",
	}

	for _, q := range questions {
		fmt.Printf("\n📝 You: %s\n", q)
		response, err := assistant.Run(ctx, q)
		if err != nil {
			log.Printf("Error: %v", err)
			continue
		}
		fmt.Printf("\n🤖 Assistant: %s\n", response)
	}
}
